cmd/devp2p/internal/ethtest/conn.go

```diff
@@ -0,0 +1,370 @@
+			return fmt.Errorf("could not negotiate snap protocol (remote caps: %v, local snap version: %v)", msg.Caps, c.ourHighestSnapProtoVersion)
+		}
+		return nil
+	default:
+		return fmt.Errorf("bad handshake: got msg code %d", code)
+	}
+}
+
+// negotiateEthProtocol sets the Conn's eth protocol version to highest
+// advertised capability from peer.
+func (c *Conn) negotiateEthProtocol(caps []p2p.Cap) {
+	var highestEthVersion uint
+	var highestSnapVersion uint
+	for _, capability := range caps {
+		switch capability.Name {
+		case "eth":
+			if capability.Version > highestEthVersion && capability.Version <= c.ourHighestProtoVersion {
+				highestEthVersion = capability.Version
+			}
+		case "snap":
+			if capability.Version > highestSnapVersion && capability.Version <= c.ourHighestSnapProtoVersion {
+				highestSnapVersion = capability.Version
+			}
+		}
+	}
+	c.negotiatedProtoVersion = highestEthVersion
+	c.negotiatedSnapProtoVersion = highestSnapVersion
+}
+
+// statusExchange performs a `Status` message exchange with the given node.
+func (c *Conn) statusExchange(chain *Chain, status *eth.StatusPacket) error {
+loop:
+	for {
+		code, data, err := c.Read()
+		if err != nil {
+			return fmt.Errorf("failed to read from connection: %w", err)
+		}
+		switch code {
+		case eth.StatusMsg + protoOffset(ethProto):
+			msg := new(eth.StatusPacket)
+			if err := rlp.DecodeBytes(data, &msg); err != nil {
+				return fmt.Errorf("error decoding status packet: %w", err)
+			}
+			if have, want := msg.Head, chain.blocks[chain.Len()-1].Hash(); have != want {
+				return fmt.Errorf("wrong head block in status, want:  %#x (block %d) have %#x",
+					want, chain.blocks[chain.Len()-1].NumberU64(), have)
+			}
+			if have, want := msg.TD.Cmp(chain.TD()), 0; have != want {
+				return fmt.Errorf("wrong TD in status: have %v want %v", have, want)
+			}
+			if have, want := msg.ForkID, chain.ForkID(); !reflect.DeepEqual(have, want) {
+				return fmt.Errorf("wrong fork ID in status: have %v, want %v", have, want)
+			}
+			if have, want := msg.ProtocolVersion, c.ourHighestProtoVersion; have != uint32(want) {
+				return fmt.Errorf("wrong protocol version: have %v, want %v", have, want)
+			}
+			// make sure eth protocol version is set for negotiation
+			if c.negotiatedProtoVersion == 0 {
+				return errors.New("eth protocol version must be set in Conn")
+			}
+			if status == nil {
+				// default status message
+				status = &eth.StatusPacket{
+					ProtocolVersion: uint32(c.negotiatedProtoVersion),
+					NetworkID:       chain.config.ChainID.Uint64(),
+					TD:              chain.TD(),
+					Head:            chain.blocks[chain.Len()-1].Hash(),
+					Genesis:         chain.blocks[0].Hash(),
+					ForkID:          chain.ForkID(),
+				}
+			}
+			if err := c.Write(ethProto, eth.StatusMsg, status); err != nil {
+				return fmt.Errorf("write to connection failed: %v", err)
+			}
+		case eth.UpgradeStatusMsg + +protoOffset(ethProto):
```

Comment:
<img width="383" alt="image" src="https://github.com/bnb-chain/bsc/assets/7995985/49ad67cd-1048-4a29-85b6-3971aba3f3ca">

strange , It can pass compile
I will modify it